components/security-mgt/org.wso2.carbon.security.mgt/src/main/java/org/wso2/carbon/security/keystore/KeyStoreAdmin.java

```diff
@@ -66,151 +72,203 @@
+                for (KeyStoreModel keyStoreModel : keyStores) {
+                    if (ALTERNATE_PRIMARY_KEYSTORE_PHANTOM_RESOURCE.equals(keyStoreModel.getFileName())) {
                         continue;
                     }
 
-                    Resource store = registry.get(ks[i]);
-                    int lastIndex = fullname.lastIndexOf("/");
-                    String name = fullname.substring(lastIndex + 1);
-                    String type = store.getProperty(SecurityConstants.PROP_TYPE);
-                    String provider = store.getProperty(SecurityConstants.PROP_PROVIDER);
-
                     KeyStoreData data = new KeyStoreData();
-                    data.setKeyStoreName(name);
-                    data.setKeyStoreType(type);
-                    data.setProvider(provider);
+                    data.setKeyStoreName(keyStoreModel.getFileName());
+                    data.setKeyStoreType(keyStoreModel.getType());
+                    data.setProvider(keyStoreModel.getProvider());
 
-                    String alias = store.getProperty(SecurityConstants.PROP_PRIVATE_KEY_ALIAS);
+                    String alias = keyStoreModel.getPrivateKeyAlias();
                     if (alias != null) {
                         data.setPrivateStore(true);
                     } else {
                         data.setPrivateStore(false);
                     }
 
-                    // Dump the generated public key to the file system for sub tenants 
                     if (!isSuperTenant) {
-                        Association[] associations = registry.getAssociations(
-                                ks[i], SecurityConstants.ASSOCIATION_TENANT_KS_PUB_KEY);
-                        if (associations != null && associations.length > 0) {
-                            Resource pubKeyResource = registry.get(associations[0].getDestinationPath());
-                            String fileName = generatePubCertFileName(ks[i],
-                                    pubKeyResource.getProperty(
-                                            SecurityConstants.PROP_TENANT_PUB_KEY_FILE_NAME_APPENDER));
-                            if (MessageContext.getCurrentMessageContext() != null) {
-                                String pubKeyFilePath = KeyStoreMgtUtil.dumpCert(
-                                        MessageContext.getCurrentMessageContext().getConfigurationContext(),
-                                        (byte[]) pubKeyResource.getContent(), fileName);
-                                data.setPubKeyFilePath(pubKeyFilePath);
+                        Optional<String> pubCertId =
+                                keyStoreDAO.getPubCertIdFromKeyStore(tenantUUID, keyStoreModel.getFileName());
+                        if (pubCertId.isPresent()) {
+                            Optional<PubCertModel> pubCert = pubCertDAO.getPubCert(pubCertId.get());
+                            if (pubCert.isPresent()) {
```

Comment:
can use flatMap in Optional

Reply:
Resolved with dc08f8317a1b35dffc5e7052c68b0da33f750149